[PyTorch] Avoid using std::regex for device string parsing in Device.cpp

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63204
* #63203

Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

dhruvbird committed:
[PyTorch] Avoid using std::regex for device string parsing in Device.cpp

Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] Avoid using std::regex for device string parsing in Device.cpp"


Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] Avoid using std::regex for device string parsing in Device.cpp"


Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] Avoid using std::regex for device string parsing in Device.cpp"


Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]

raziel commented on c10/core/Device.cpp:
```diff
@@ -65,33 +49,84 @@ DeviceType parse_type(const std::string& device_string) {
+  DeviceStringParsingState pstate = DeviceStringParsingState::START;
+
+  // The code below tries to match the string in the variable
+  // device_string against thr regular expression:
```
```suggestion
  // device_string against the regular expression:
```

raziel commented on c10/core/Device.cpp:
```diff
@@ -65,33 +49,84 @@ DeviceType parse_type(const std::string& device_string) {
-          "' in device string '",
-          device_string,
-          "'");
+  std::string device_name, device_index_str;
```
what about extracting this as a function and have a unit test?

dhruvbird replied:
Good point - I initially thought of doing that (extract into a separate function) but then realized that the function's body would be exactly the same as this overloaded constructor, so skipped that step. Added unit tests for this constructor here: https://www.internalfb.com/diff/D30286910

dhruvbird replied:
GH PR link: https://github.com/pytorch/pytorch/pull/63203 (for the unit test).

raziel commented on c10/core/Device.cpp:
```diff
@@ -6,26 +6,10 @@
 #include <array>
 #include <exception>
```
don't you need to add #include <ctype.h> since you use isalpha?

dhruvbird replied:
Ah! that's an oversight - thanks!

dhruvbird replied:
Strangely it worked w/o the header - probably because of transitive includes, but better to include it directly.

raziel commented on c10/core/Device.cpp:
```diff
@@ -65,33 +49,84 @@ DeviceType parse_type(const std::string& device_string) {
+            pstate = DeviceStringParsingState::ERROR;
+          }
+        } else {
+          pstate = DeviceStringParsingState::INDEX_START;
```
It seems you are not enforcing the + in ([a-zA-Z_]+)
Basically, you can start from "::" without error.

dhruvbird replied:
ah! Good observation! This is bein enforced later in the code here:

```
  const bool has_error = device_name.empty() ||
      pstate == DeviceStringParsingState::ERROR ||
      (pstate == DeviceStringParsingState::INDEX_START &&
       device_index_str.empty());

  TORCH_CHECK(!has_error, "Invalid device string: '", device_string, "'");
```

Basically I wanted just a single exception throw for all cases that the previous regex would throw on to keep semantics and also not have multiple places where an exception is thrown. I also added this test case in the previous diff just to make sure that we continue to discard the faulty string in the future.

raziel replied:
Ah got it. Thanks!

raziel reviewed (changes requested):
Personally I would have written the code to split the regex in sections (i.e. 3 while loops), rather than using the switch and a state for them, but that's just me :-)

Anyway, I have some comments.

dhruvbird committed:
Update on "[PyTorch] Avoid using std::regex for device string parsing in Device.cpp"


Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] Avoid using std::regex for device string parsing in Device.cpp"


Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] Avoid using std::regex for device string parsing in Device.cpp"


Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]

raziel reviewed (approved):
Thanks!

dhruvbird committed:
Update on "[PyTorch] Avoid using std::regex for device string parsing in Device.cpp"


Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] Avoid using std::regex for device string parsing in Device.cpp"


Currently, `std::regex` is used for parsing device strings. This is undesirable for a few reasons.

1. Increases binary size
2. Slows down model loading
3. Potentially uses more memory at runtime
4. Takes marginally longer time to build code that uses std::regex v/s not using std::regex

This change avoids the use of `std::regex` for parsing the device string since we don't need to.

Differential Revision: [D30281388](https://our.internmc.facebook.com/intern/diff/D30281388/)

[ghstack-poisoned]
